syntax = "proto3";

option go_package = "protobuf/api";

package api;

service API {
    rpc GetModelParams(Empty) returns (ModelParameters) {}
    rpc ReturnModel(ModelResults) returns (Empty) {}
}

message Empty {}

enum Optimizer {
    Adam = 0;
    SGD = 1;
    RMSprop = 2;
}

enum ActivationFunc {
    Relu = 0;
    Sigmoid = 1;
    Softmax = 2;
    Tanh = 3;
}

message Layer {
    int32 num_neurons = 1;
}

message ModelParameters {
    string model_id = 1;

    float learning_rate = 2;
    Optimizer optimizer = 3;
    ActivationFunc activation_func = 4;
    repeated Layer layers = 5;
    bool dropout = 6;
}

message ModelResults {
    string model_id = 1;

    float recall = 2;
}